Mods can disable user's posting, commenting and PMing privileges.

Raemon opened the pull request:
This PR moves us towards a world where we can selectively ban users from certain activities without completely banning them from the site. Currently, on LW when someone starts making a lot of mediocre posts we sometimes ban them, but this feels pretty sad since it means they can't keep using the logged in read-status for posts, or Continue Reading, etc.

I think it's fairly likely we'll want to ban some people from making top-level posts but not necessarily comments, so it sets up banning Posts/Comments/PMs separately. 

This post doesn't currently enable banning users from managing tags because that was kind of complicated, but it seemed worth shipping these features for now.

darkruby501 commented:
Tests are failing

darkruby501 commented on packages/lesswrong/lib/collections/users/fragments.ts:
```diff
@@ -320,6 +323,8 @@ registerFragment(`
     deleteContent
     banned
 
+    # banned from individual actions
```
What's going on here?

darkruby501 commented on packages/lesswrong/lib/collections/users/fragments.ts:
```diff
@@ -81,6 +81,9 @@ registerFragment(`
+    bannedFromPosting
+    bannedFromCommenting
+    bannedFromPMing
   }
```
What about banning from voting? Not sure whether that should be included or not. I do think banned users are people whose judgment we might dislike and who might do more retalitory stuff

Raemon replied:
I do think we should have banning from voting but voting was a bit of a weirder system to interact with, so was kinda punting it to the future.

Raemon replied:
(part of my decisionmaking here is "this isn't the top priority so I wanted to do stuff that was a simple improvement that didn't require much effort to reason about.")

darkruby501 reviewed (changes requested):
Obviously fix the failing tests. Other thing is ensure that actual permissions are restricted, not just UI components. This probably means affecting the collection permissions. `userCanDo` stuff

Raemon commented:
(this PR deliberately avoids the scope of "decide when to do the disabling" because that turned out to be tricky/controversial)

Raemon commented on packages/lesswrong/components/users/UsersProfile.tsx:
```diff
@@ -503,6 +504,8 @@ const UsersProfileFn = ({terms, slug, classes}: {
       </a>}
     </>
 
+    const showMessageButton = currentUser?._id != user._id && userCanStartConversations(currentUser)
```
I originally was going to block all new messages, and then realized it was actually quite weird to disable a user's ability to reply to existing conversations. For example, if the moderators sent them a message they wouldn't be able to reply. The main probably I was thinking I might want to solve is stopping users from spamming other users, which is mostly. new-conversation thing.

Agreed about removing the message button when it appears other places though.

Raemon commented on packages/lesswrong/lib/collections/users/helpers.ts:
```diff
@@ -161,17 +161,11 @@ export const userIsBannedFromAllPersonalPosts = (user: UsersCurrent|DbUser, post
 }
 
 export const userIsAllowedToComment = (user: UsersCurrent|DbUser|null, post: PostsDetails|DbPost, postAuthor: PostsAuthors_user|DbUser|null): boolean => {
```
Yup that sounds right. Thanks!

Raemon replied:
(I'm probably going to fix this in a separate PR)

Raemon commented on packages/lesswrong/lib/collections/users/custom_fields.ts:
```diff
@@ -1788,6 +1788,36 @@ addFieldsDict(Users, {
+    group: formGroups.disabledPrivileges,
+    order: 69,
+  },
+  commentingDisabled: {
```
I do think it's quite plausible that "only allow people to comment on their shortform" is a setting we'd want. (This might actually be the main way we want mostly-banned-users to continue to contribute, since it neither clogs up Latest Posts or the comment threads of other posts)

Right now this PR isn't top priority so I'd prefer to ship something close to the current version for now, but I agree that's a pretty reasonable tweak to it.

Raemon commented on packages/lesswrong/components/posts/PostsNewForm.tsx:
```diff
@@ -130,9 +130,16 @@ const PostsNewForm = ({classes}: {
+  if (!currentUser) {
     return (<WrappedLoginForm />);
   }
+  if (!userCanPost(currentUser)) {
+    return (<SingleColumnSection>
+      <Typography variant="display1">
+        You don't have permission to post
+      </Typography>
+    </SingleColumnSection>);
+  }
```
I weakly lean towards "fine for users to publish or edit existing drafts/posts". If it turns out to be a problem we can fix that later.

Raemon commented:
Additional changes:

Separating out commentingDisabled, and commentingOnOtherUsersDisabled. Much of the time I think it's fine for users to comment on their own posts (including shortform)

Removed a bunch of additional UI elements (various New Shortforms, other types of new posts)

Moved the logic for hiding the Message button into the NewConversationButton, so it's hidden wherever we use it.

darkruby501 commented on packages/lesswrong/lib/collections/users/helpers.ts:
```diff
@@ -161,17 +161,12 @@ export const userIsBannedFromAllPersonalPosts = (user: UsersCurrent|DbUser, post
+  if (!user) return false
+  if (user.deleted) return false
+  if (user.allCommentingDisabled) return false
+  if (user.commentingOnOtherUsersDisabled && (post.userId && (post.userId != user._id))) return false // this has to check for post.userId because that isn't provided to shorform components, which are meant to be allowed
```
hmm, actually overall let's figure out this line
